src/sentry/tasks/assemble.py

```diff
@@ -61,19 +130,162 @@ def assemble_dif(project_id, name, checksum, chunks, **kwargs):
+        logger.error('failed to assemble dif', exc_info=True)
+    else:
+        set_assemble_status(AssembleTask.DIF, project, checksum, ChunkFileState.OK,
+                            detail=serialize(dif))
     finally:
         if delete_file:
             file.delete()
 
 
-def assemble_file(project, name, checksum, chunks, file_type):
-    '''This assembles multiple chunks into on File.'''
-    from sentry.models import File, ChunkFileState, AssembleChecksumMismatch, \
-        FileBlob, set_assemble_status
+class AssembleArtifactsError(Exception):
+    pass
+
+
+@instrumented_task(name='sentry.tasks.assemble.assemble_artifacts', queue='assemble')
+def assemble_artifacts(org_id, version, checksum, chunks, **kwargs):
+    """
+    Creates release files from an uploaded artifact bundle.
+    """
+
+    import shutil
+    import tempfile
+    from sentry.utils.zip import safe_extract_zip
+    from sentry.models import File, Organization, Release, ReleaseFile
+
+    with configure_scope() as scope:
+        scope.set_tag("organization", org_id)
+
+    organization = Organization.objects.filter(id=org_id).get()
+    set_assemble_status(AssembleTask.ARTIFACTS, organization, checksum, ChunkFileState.ASSEMBLING)
+
+    # Assemble the chunks into a temporary file
+    rv = assemble_file(AssembleTask.ARTIFACTS, organization, 'release-artifacts.zip',
+                       checksum, chunks, file_type='release.bundle')
+
+    # If not file has been created this means that the file failed to
+    # assemble because of bad input data.  Return.
+    if rv is None:
+        return
+
+    bundle, temp_file = rv
+    scratchpad = tempfile.mkdtemp()
+
+    # Initially, always delete the bundle file. Later on, we can start to store
+    # the artifact bundle as a release file.
+    delete_bundle = True
+
+    try:
+        try:
+            safe_extract_zip(temp_file, scratchpad, strip_toplevel=False)
+        except BaseException:
+            raise AssembleArtifactsError('failed to extract bundle')
+
+        try:
+            manifest_path = path.join(scratchpad, 'manifest.json')
+            with open(manifest_path, 'rb') as manifest:
+                manifest = json.loads(manifest.read())
+        except BaseException:
+            raise AssembleArtifactsError('failed to open release manifest')
+
+        org_slug = manifest.get('org')
+        if organization.slug != org_slug:
+            raise AssembleArtifactsError('organization does not match uploaded bundle')
+
+        release_name = manifest.get('release')
+        if release_name != version:
+            raise AssembleArtifactsError('release does not match uploaded bundle')
+
+        try:
+            release = Release.objects.get(
+                organization_id=organization.id,
+                version=release_name
+            )
+        except Release.DoesNotExist:
+            raise AssembleArtifactsError('release does not exist')
```

Comment:
Should we not upsert the release here like we do elsewhere?

Reply:
Ah, you mean `get_or_create`? Yeah, I guess that could work. At first I had a different structure. Will change 👍

Reply:
Ah, just realizing that I misread this comment. We do not upsert releases in other places of the API either. However, `sentry-cli` issues the `releases new` command before uploading files. See [here](https://github.com/getsentry/sentry/blob/2280925c5564756234fda627c7999ba75de3241f/src/sentry/api/endpoints/organization_release_files.py#L108-L115) for reference.